cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/LineageOperationsProcessor.java

```diff
@@ -173,9 +175,32 @@ private void addMergeOperation(Set<String> stageInputs,
                                              Map<String, Operation> processedOperations) {
     // We need to return InputFields in the same order as fields.
     // Keep them in map so we can iterate later on received fields to return InputFields.
+
     Map<String, InputField> inputFields = new HashMap<>();
     List<String> parents = findParentStages(currentStage);
+
+    List<String> fieldsForJoin = new ArrayList<>();
     for (String field : fields) {
+      if (noMergeRequiredStages.contains(currentStage)) {
+        // Current stage is of type JOIN.
+        // JOIN creates operation with input fields in format <stagename.fieldname>
+        // if the field is directly read from the schema of the one of the input stage.
+        // However it is also possible to have operation with input not tagged with the stage name, for example
+        // in case if rename is done on the join keys. In this case Join operation will read fields of the form
+        // <stagename.fieldname> however rename operations followed by it will simply read the output of the
+        // Join which wont be tagged with stagename.
+        Iterable<String> stageFieldPair = Splitter.on(".").omitEmptyStrings().trimResults().split(field);
```

Comment:
Stage names and field names are both allowed to have '.' in them. There is no way to do this correctly with a separator. Is there a jira open to do join lineage properly?  If so, can you add a TODO with the jira number. If not, can you open a jira and add a TODO with the number.

A slightly better, but still bad option is to use an uncommon character as a separator.

Another slightly better option is to check for the entire stage name plus a '.' as a prefix, and use that to split into stage name and field name.

Either way there are still scenarios where things will be buggy, such as when a field name happens to match the stage name.

Reply:
I see. Added more stricter check that prefix is a part of stage name per suggestion. Also added a TODO (Join API).

On a separate note, I implemented this logic based on the current `Joiner` plugin which asks user to specify join keys and aliases as `StageName.fieldname`. So if either of stage name or field name has `.` in it then the plugin won't work right?

Reply:
If the Joiner is using '.' to split, it is also wrong. It should be using the input stages as the prefixes. Over there, it is unambiguous because it expects everything will be prefixed by stage name. Here, even without the delimiter problem, it can be ambiguous if there is a field name that matches the stage name.